latent_programmer/train_test.py

```diff
@@ -0,0 +1,191 @@
+# limitations under the License.
+
+# python3
+"""Test seq-to-seq training."""
+
+# pytype: disable=wrong-arg-count
+# pytype: disable=attribute-error
+
+import collections
+import functools
+import json
+import os
+import sys
+sys.path.append('..')
+import random
+import time
+
+from absl.testing import absltest
+
+from flax import jax_utils
+from flax import linen as nn
+from flax import optim
+from flax.training import common_utils
+import jax
+import jax.numpy as jnp
+import numpy as np
+import tensorflow.compat.v2 as tf
+
+from latent_programmer import decode
+from latent_programmer import models
+from latent_programmer import train_lib
+from latent_programmer.tasks.robust_fill import dsl
+from latent_programmer.tasks.robust_fill import tokens as dsl_tokens
+from latent_programmer.tasks.robust_fill.dataset import input_pipeline
+
+gfile = tf.io.gfile
+
+class TrainTest(absltest.TestCase):
+
+  def test_train(self):
+    tf.enable_v2_behavior()
+
+    tf.random.set_seed(0)
+    np.random.seed(0)
+    random.seed(0)
+
+    dataset_filepattern = 'tasks/robust_fill/dataset/test_dataset/program_tasks.tf_records-*'
+
+    batch_size = 4
+    num_strings_per_task = 4
+    max_characters = 10
+    max_program_length = 15
+
+    # Build token tables.
+    id_char_table = {i+1: char for (i, char) in enumerate(dsl.CHARACTER)}
+    char_id_table = {char: id for id, char in id_char_table.items()}
+    id_token_table, token_id_table = dsl_tokens.build_token_tables()
+    io_vocab_size = len(char_id_table) + 1  # For padding.
+    program_vocab_size = len(token_id_table) + 1
+
+    bos_token = token_id_table[dsl.BOS]
+    eos_token = token_id_table[dsl.EOS]
+
+    def decode_io(inputs, outputs):
+      """Decode io examples tokens."""
+      def decode_str(s):
+        """Decode string tokens."""
+        return ''.join([id_char_table[c_id] for c_id in s if c_id > 0])
+
+      io_string = ''
+      inps, outs = [], []
+      for inp, out in zip(inputs, outputs):
+        inps.append(decode_str(inp))
+        outs.append(decode_str(out))
+        io_string += inps[-1] + ' < ' + outs[-1] + ' > '
+      return inps, outs
+
+    def decode_program(program):
+      """Decode program tokens."""
+      program = program[:np.argmax(program == eos_token) + 1].astype(np.int32)
+      try:
+        p = dsl.decode_program(program, id_token_table)
+        return p
+      except:  # pylint: disable=bare-except
+        return None
+
+    # Load dataset.
+    dataset = input_pipeline.create_dataset_from_tf_record(
+      dataset_filepattern, token_id_table, char_id_table)
+    dataset = dataset.padded_batch(
+      batch_size,
+      padded_shapes=((num_strings_per_task, max_characters),
+                     (num_strings_per_task, max_characters),
+                     (max_program_length,)),
+      drop_remainder=True)
+    dataset_iter = dataset.repeat().as_numpy_iterator()
+
+    train_config = models.TransformerConfig(
```

Comment:
This model isn't super small - how long does it take to run this test? let's aim for at most a couple minutes. also the test machines have 2 cores and 4GB memory, if that helps to calibrate